File: php/Generator/Generator.php
<?php

interface Builder {
    function reset();
    function setWheels($num);
    function setSeats($num);
    function setEngine($obj);
}

class CarBuilder implements Builder {

    private $product;

    function __construct()
    {
        $this->reset();

    }

    function reset()
    {
        $this->product = new Car();
    }


    function setWheels($num)
    {
        $this->product->wheel = "{$num} car wheel(s)";
    }

    function setSeats($num)
    {
        $this->product->seat = "{$num} car seat(s)";
    }

    function setEngine($obj)
    {
        $this->product->engine = "car engine is {$obj}";
    }

    function getProduct()
    {
        $product =  $this->product;
        $this->reset();
        return $product;
    }
}

class BicycleBuilder implements Builder {
    private $product;
    function __construct()
    {
        $this->reset();
    }

    function reset()
    {
        $this->product = new Bicycle();
    }

    function setWheels($num)
    {
        $this->product->wheel = "{$num} bicycle wheel(s)";
    }

    function setSeats($num)
    {
        $this->product->seat = "{$num} bicycle seat(s)";
    }

    function setEngine($obj)
    {
        $this->product->engine = "bicycle engine is {$obj}";
    }

    function getProduct()
    {
        $product = $this->product;
        $this->reset();
        return $product;
    }
}

class Car {
    public $engine;
    public $seat;
    public $wheel;
}

class Bicycle {
    public $wheel;
    public $seat;
    public $engine;
}

class Director {
    public static function buildNewCar(Builder $builder) {
        $builder->reset();
        //
        $builder->setWheels(4);
        $builder->setSeats(2);  //maybe more
        $builder->setEngine("v8");
    }

    public static function buildBicycle(Builder $builder) {
        $builder->reset();
        $builder->setWheels(2);
        $builder->setSeats(1); //
        $builder->setEngine("manual");
    }
}

$builder = new CarBuilder();
Director::buildNewCar($builder);
$car = $builder->getProduct();
var_dump($car->wheel);
var_dump($car->seat);
var_dump($car->engine);

File: go/src/Generator/generator.go
package Generator

import "fmt"

type Car struct {
	wheel  int
	seat   int
	engine string
}

type Bicycle struct {
	wheel  int
	seat   int
	engine string
}

type Builder interface {
	setSeat(num int)
	setWheel(num int)
	setEngine(obj string)
	reset()
}

type CarBuilder struct {
	Product *Car
}
type BicycleBuilder struct {
	Product *Bicycle
}

func (c *CarBuilder) getProduct() *Car {
	product := c.Product
	c.reset()
	return product
}

func (c *CarBuilder) reset() {
	c.Product = new(Car)
}

func (c *CarBuilder) setSeat(num int) {
	c.Product.seat = num
}

func (c *CarBuilder) setWheel(num int) {
	c.Product.wheel = num
}

func (c *CarBuilder) setEngine(obj string) {
	c.Product.engine = obj
}

func (b *BicycleBuilder) setSeat(num int) {
	b.Product.seat = num
}
func (b *BicycleBuilder) setWheel(num int) {
	b.Product.wheel = num
}

func (b *BicycleBuilder) setEngine(obj string) {
	b.Product.engine = obj
}

func (b *BicycleBuilder) reset() {
	b.Product = new(Bicycle)
}

func (b *BicycleBuilder) getProduct() *Bicycle {
	product := b.Product
	b.reset()
	return product
}

func Execute() {
	builder := &BicycleBuilder{}
	//the following steps can be setted in a director class
	builder.reset()
	builder.setWheel(2)
	builder.setSeat(1)
	builder.setEngine("manual")

	bicycle := builder.getProduct()
	fmt.Println(bicycle.seat)
	fmt.Println(bicycle.wheel)
	fmt.Println(bicycle.engine)

}
